source/opt/loop_descriptor.cpp

```diff
@@ -0,0 +1,162 @@
+           BasicBlock* merge_target)
+    : loop_header_(header),
+      loop_continue_(continue_target),
+      loop_merge_(merge_target),
+      loop_preheader_(nullptr),
+      parent_(nullptr) {
+  assert(context);
+  assert(dom_analysis);
+  loop_preheader_ = FindLoopPreheader(context, dom_analysis);
+  AddBasicBlockToLoop(header);
+  AddBasicBlockToLoop(continue_target);
+}
+
+BasicBlock* Loop::FindLoopPreheader(IRContext* ir_context,
+                                    opt::DominatorAnalysis* dom_analysis) {
+  CFG* cfg = ir_context->cfg();
+  opt::DominatorTree& dom_tree = dom_analysis->GetDomTree();
+  opt::DominatorTreeNode* header_node = dom_tree.GetTreeNode(loop_header_);
+
+  // The loop predecessor.
+  BasicBlock* loop_pred = nullptr;
+
+  auto header_pred = cfg->preds(loop_header_->id());
+  for (uint32_t p_id : header_pred) {
+    opt::DominatorTreeNode* node = dom_tree.GetTreeNode(p_id);
+    if (node && !dom_tree.Dominates(header_node, node)) {
+      // The predecessor is not part of the loop, so potential loop preheader.
+      if (loop_pred && node->bb_ != loop_pred) {
+        // If we saw 2 distinct predecessors that are outside the loop, we don't
+        // have a loop preheader.
+        return nullptr;
+      }
+      loop_pred = node->bb_;
+    }
+  }
+  // Safe guard against invalid code, SPIR-V spec forbids loop with the entry
+  // node as header.
+  assert(loop_pred && "The header node is the entry block ?");
+
+  // So we have a unique basic block that can enter this loop.
+  // If this loop is the unique successor of this block, then it is a loop
+  // preheader.
+  bool is_preheader = true;
+  uint32_t loop_header_id = loop_header_->id();
+  loop_pred->ForEachSuccessorLabel(
+      [&is_preheader, loop_header_id](const uint32_t id) {
+        if (id != loop_header_id) is_preheader = false;
+      });
+  if (is_preheader) return loop_pred;
+  return nullptr;
+}
+
+LoopDescriptor::LoopDescriptor(const Function* f) { PopulateList(f); }
+
+void LoopDescriptor::PopulateList(const Function* f) {
+  IRContext* context = f->GetParent()->context();
+
+  opt::DominatorAnalysis* dom_analysis =
+      context->GetDominatorAnalysis(f, *context->cfg());
+
+  loops_.clear();
+
+  // Traverse the tree and apply the above functor to find all the OpLoopMerge
+  // instructions. Instructions will be in domination order of BasicBlocks.
+  // However, this does not mean that dominance is implied by the order of
+  // loop_merge_inst you still need to check dominance between each block
+  // manually.
+  opt::DominatorTree& dom_tree = dom_analysis->GetDomTree();
+  // Post-order traversal of the dominator tree: inner loop will be inserted
+  // first.
+  for (opt::DominatorTreeNode& node :
+       ir::make_range(dom_tree.post_begin(), dom_tree.post_end())) {
+    Instruction* merge_inst = node.bb_->GetLoopMergeInst();
+    if (merge_inst) {
+      // The id of the continue basic block of this loop.
```

Comment:
I think you mean the merge basic block, not the continue basic block.

Reply:
Indeed, thanks.